pipeline.cpp

```diff
@@ -1,200 +1,237 @@
+{
+    SDL_UpdateTexture(GPU_OUTPUT, NULL, frameBuf->pixels, frameBuf->pitch);
+    SDL_RenderClear(ren);
+    SDL_RenderCopy(ren, GPU_OUTPUT, NULL, NULL);
+    SDL_RenderPresent(ren);
+}
+
+/*************************************************************
+ * POLL_CONTROLS
+ * Updates the state of the application based on:
+ * keyboard, mouse, touch screen, gamepad inputs. 
+ ************************************************************/
+void processUserInputs(bool & running)
+{
+    SDL_Event e;
+    int mouseX;
+    int mouseY;
+    while(SDL_PollEvent(&e)) 
+    {
+        if(e.type == SDL_QUIT) 
+        {
+            running = false;
+        }
+        if(e.key.keysym.sym == 'q' && e.type == SDL_KEYDOWN) 
+        {
+            running = false;
+        }
+    }
+}
+
+/****************************************
+ * DRAW_POINT
+ * Renders a point to the screen with the
+ * appropriate coloring.
+ ***************************************/
+void DrawPoint(Buffer2D<PIXEL> & target, Vertex* v, Attributes* attrs, Attributes * const uniforms, FragmentShader* const frag)
+{
+    // Your code goes here
+	target[(int)v[0].y][(int)v[0].x] = attrs[0].color;
+}
+
+/****************************************
+ * DRAW_TRIANGLE
+ * Renders a line to the screen.
+ ***************************************/
+void DrawLine(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* const attrs, Attributes* const uniforms, FragmentShader* const frag)
+{
+    // Your code goes here
+}
+
+/*************************************************************
+ * DRAW_TRIANGLE
+ * Renders a triangle to the target buffer. Essential 
+ * building block for most of drawing.
+ ************************************************************/
+void DrawTriangle(Buffer2D<PIXEL> & target, Vertex* const triangle, Attributes* const attrs, Attributes* const uniforms, FragmentShader* const frag)
+{
+    // Your code goes here
+    //the min and max, x and y coordinates for the triangles, which would create a rectangle 
+    int maxX = MAX3(triangle[0].x, triangle[1].x, triangle[2].x);
+    int maxY = MAX3(triangle[0].y, triangle[1].y, triangle[2].y);
+    int minX = MIN3(triangle[0].x, triangle[1].x, triangle[2].x);
+    int minY = MIN3(triangle[0].y, triangle[1].y, triangle[2].y);
+
+    //the x and y coordinates for each triangle
+    Vertex a = {(float)triangle[0].x, (float)triangle[0].y};
+    Vertex b = {(float)triangle[1].x, (float)triangle[1].y};
+    Vertex c = {(float)triangle[2].x, (float)triangle[2].y};
+
+    //deciding which points of the square are in the triangle
+    for (int y = minY; y <= maxY; y++)
+    {
+        for (int x = minX; x <= maxX; x++)
+        {
+            //the point to fill in or not
+            Vertex v = {(float)x, (float)y};
+
+            //calculating the barycentric coordinates
+            float baryA = (((b.y - c.y)*(v.x - c.x))+((c.x - b.x)*(v.y - c.y))) / (((b.y - c.y)*(a.x - c.x))+((c.x - b.x)*(a.y - c.y)));

```

Comment:
I wrote a separate cross product function for this section but this is probably just as efficient.